### PR TITLE
[1.0.1] Verify block signature first when creating block_state

### PR DESCRIPTION
**Pull request by heifner**
When creating a `block_state`, verify the block signature before ingesting anything from the block.

### CODE REVIEW

**Comment by greg7mdp** [comment]
Shouldn't we verify the block signature even before we start validating the block header extensions?

**Comment by heifner** [comment]
> Shouldn't we verify the block signature even before we start validating the block header extensions? First thing in `create_block_state_i`?

I think in general we should consider verification to be allowed in parallel. The point of this PR is that we complete block signature verification before integrating the block. The order of any verification before integration shouldn't matter.

**Comment by greg7mdp** [comment]
> > Shouldn't we verify the block signature even before we start validating the block header extensions? First thing in `create_block_state_i`?
> 
> I think in general we should consider verification to be allowed in parallel. The point of this PR is that we complete block signature verification before integrating the block. The order of any verification before integration shouldn't matter.

In general I agree, except that our extension/qc verification code itself could have bugs causing nodeos to abort. It seems safer to be to validate signatures first (old reliable code), to ensure that the block header extensions that we verify next were not tampered with.

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: When creating a `block_state`, verify the block signature before ingesting anything from the block..
Note:end

File: libraries/chain/block_state.cpp [hunk 1]
```diff
@@ -9,20 +9,66 @@
...
+
+} // namespace detail
+
+using namespace eosio::chain::detail;
+
+// ASSUMPTION FROM controller_impl::apply_block = all untrusted blocks will have their signatures pre-validated here
 block_state::block_state(const block_header_state& prev, signed_block_ptr b, const protocol_feature_set& pfs,
                          const validator_t& validator, bool skip_validate_signee)
-   : block_header_state(prev.next(*b, validator))
+   : block_header_state(verify_block_sig(prev, b, skip_validate_signee) ? prev.next(*b, validator) : block_header_state{})
```
Comment (greg7mdp) [inline libraries/chain/block_state.cpp:66]: `verify_block_sig` always returns true, so the ternary operator is weird. I think the following is clearer maybe:
```suggestion
   : block_header_state(verify_block_sig(prev, b, skip_validate_signee), prev.next(*b, validator))
```

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/block_state.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/block_state.cpp]: I tried that first, but compiler doesn't like it:
```
/home/heifner/ext/leap/libraries/chain/block_state.cpp:66:25: error: no viable conversion from 'bool' to 'block_id_type' (aka 'fc::sha256')
   : block_header_state(verify_block_sig(prev, b, skip_validate_signee), prev.next(*b, validator))
```
However, an extra set of `()` seems to do the trick.

**Review by linh2931 — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
category: Other
component: Internal
summary: When creating a block_state, verify the block signature before ingesting anything from the block.
Note:end

**Pull request merged**